### PR TITLE
feat: add download completed notification to settings

### LINKED ISSUE #234: Notice when finishing downloading
**Describe the solution you'd like**
I notice that there's a notice when I start a download from browser with surge extension. Could you add such a notice when finishing downloading? This notice may contain a button to open the download directory or the download file directly.

### PR DESCRIPTION
**Pull request by SuperCoolPencil**
Close #234 

image

<h3>Greptile Summary</h3>

This PR adds a **Download Complete Notification** setting that fires a system notification (via `github.com/gen2brain/beeep`) on download success, finalization failure, and download error. The setting is correctly plumbed through `GeneralSettings`, exposed in the TUI settings view, and defaults to `true`. Four new tests cover suppression, the zero-elapsed edge case, and the DownloadID fallback.

Key issues found:

- **Dead code in the finalization-failure notification block** (`events.go` line 235): `msg := "Download failed"` is immediately overwritten because the surrounding `if err != nil` guarantees `err` is non-nil — the default string is never reachable.
- **Missing `testutil.SetupStateDB` in `TestStartEventWorker_ErrorNotificationFallsBackToDownloadID`** (`events_internal_test.go`): all sibling tests that touch state call this helper; omitting it risks leaking or reading stale DB state in parallel test runs.
- **Duplicate anonymous-struct notify helper** repeated across three test functions; a small shared `captureNotify` helper would remove the boilerplate and make the capture type easier to evolve.
- **Spurious blank line** at the top of `utils.Notify`'s body will be flagged by `golangci-lint`.
- The new `binary-size-compare.yml` workflow uses an empty `signature` string which means `body.includes("")` always matches the very first PR comment, causing the bot to overwrite an unrelated commenter's post on subsequent workflow runs.

<h3>Confidence Score: 2/5</h3>

- Not safe to merge as-is — the binary-size workflow has a comment-overwrite bug affecting all PRs, the dead-code logic is misleading, and the test isolation gap could hide real failures.
- The core feature is implemented correctly and most tests are thorough, but the empty-signature bug in the new CI workflow will silently corrupt unrelated PR comments for every contributor, the dead-code `if err != nil` check in the finalization failure path misrepresents control flow, and the missing `testutil.SetupStateDB` call leaves one test without proper isolation. These issues collectively lower confidence before merge.
- `.github/workflows/binary-size-compare.yml` (empty signature), `internal/processing/events.go` (dead-code msg default), `internal/processing/events_internal_test.go` (missing SetupStateDB + duplicate helpers)

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| internal/processing/events.go | Adds three notification blocks (finalization failure, success, download error); contains dead-code `if err != nil` guard in the finalization failure path and duplicate filename-resolution/notify patterns across all three blocks. |
| internal/processing/events_internal_test.go | Adds four new notification tests with good coverage of happy path, suppression, and fallback; `TestStartEventWorker_ErrorNotificationFallsBackToDownloadID` is missing `testutil.SetupStateDB` and all three capturing tests duplicate the same anonymous-struct notify helper. |
| internal/utils/notify.go | New utility wrapping `beeep.Notify`; correctly guards the icon write with an existence check; has a spurious blank line at the top of `Notify` that will fail lint. |
| internal/config/settings.go | Clean addition of `DownloadCompleteNotification` bool field with metadata entry and default value of `true`; no issues found. |
| .github/workflows/binary-size-compare.yml | New workflow for binary-size comparison; the empty `signature` string causes `body.includes("")` to always match the first PR comment rather than the bot's own, corrupting unrelated comments on update runs. |
| go.mod | Adds `github.com/gen2brain/beeep` with 8 transitive indirect dependencies including image-processing and systray packages unrelated to text notifications, meaningfully increasing binary size. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant W as EventWorker
    participant S as Settings
    participant U as utils.Notify
    participant B as beeep

    W->>W: Receive DownloadCompleteMsg
    W->>W: finalizeCompletedFile(destPath)
    alt Finalization fails
        W->>S: GetSettings()
        S-->>W: settings
        alt DownloadCompleteNotification == true
            W->>U: Notify("Download failed: <filename>", err.Error())
            U->>B: beeep.Notify(title, msg, iconPath)
        end
    else Finalization succeeds
        W->>S: GetSettings()
        S-->>W: settings
        alt DownloadCompleteNotification == true
            alt Elapsed > 0
                W->>U: Notify("Download Complete: <filename>", "…speed…")
            else Elapsed == 0
                W->>U: Notify("Download Complete: <filename>", "Download complete!")
            end
            U->>B: beeep.Notify(title, msg, iconPath)
        end
    end

    W->>W: Receive DownloadErrorMsg
    W->>S: GetSettings()
    S-->>W: settings
    alt DownloadCompleteNotification == true
        W->>U: Notify("Download failed: <filename|id>", err.Error())
        U->>B: beeep.Notify(title, msg, iconPath)
    end
```

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `internal/processing/events.go`, line 295-306 ([link](https://github.com/surge-downloader/surge/blob/8c57313f057ab95e1c35fab76427e59575dd616c/internal/processing/events.go#L295-L306)) 

   <a href="#">P1</a> **Error notification ignores user setting**

   The `DownloadErrorMsg` notification is also unconditional. The same `DownloadCompleteNotification` flag that controls the success path should control this path too. Additionally, the title uses `"download failed:"` with a lowercase "d", inconsistent with `"Download failed:"` used elsewhere.

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/processing/events.go
   Line: 295-306

   Comment:
   **Error notification ignores user setting**

   The `DownloadErrorMsg` notification is also unconditional. The same `DownloadCompleteNotification` flag that controls the success path should control this path too. Additionally, the title uses `"download failed:"` with a lowercase "d", inconsistent with `"Download failed:"` used elsewhere.

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: internal/processing/events.go
Line: 235-238

Comment:
**Dead default — `"Download failed"` string is unreachable**

This code is nested inside `if err := finalizeCompletedFile(destPath); err != nil { … }`, so `err` is guaranteed to be non-nil at this point. The inner `if err != nil` is always `true`, which means the default `msg := "Download failed"` can never survive to the `notify` call — `msg` is always overwritten by `err.Error()`. The fallback string is dead code that misleads readers into thinking there's a nil-error path here.

```suggestion
				msg := err.Error()
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/processing/events_internal_test.go
Line: 651-656

Comment:
**Missing `testutil.SetupStateDB` call — inconsistent test isolation**

Every other test in this file that exercises a `DownloadErrorMsg` or `DownloadCompleteMsg` path (including the just-above `TestStartEventWorker_RemovesIncompleteFileOnErrorWithoutDBEntry`) begins with `testutil.SetupStateDB(t)` to ensure the state DB is pointing at an isolated temp directory. This test omits that call, so it can silently read from (or corrupt) whatever global DB state has leaked from a previous parallel test. If the test is relying on `state.GetDownload` returning `nil` due to an uninitialised DB, that behaviour should be made explicit, not relied upon incidentally.

```suggestion
func TestStartEventWorker_ErrorNotificationFallsBackToDownloadID(t *testing.T) {
	testutil.SetupStateDB(t)

	settingsDir := t.TempDir()
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/processing/events_internal_test.go
Line: 494-503

Comment:
**Repeated anonymous struct across three test functions**

The same anonymous struct type (`struct { title string; msg string }`) and the same mock-notify closure are copy-pasted verbatim into `TestStartEventWorker_MarksCompletionAsErrorWhenFinalizationFails` (here), `TestStartEventWorker_CompletionNotificationUsesGenericMessageWhenElapsedZero` (line ~616), and `TestStartEventWorker_ErrorNotificationFallsBackToDownloadID` (line ~661). This violates the no-duplicate-logic guideline and makes future changes to the capture signature error-prone.

Consider extracting a shared helper at the top of the test file:

```go
type notifyCall struct {
    title string
    msg   string
}

func captureNotify(t *testing.T) *[]notifyCall {
    t.Helper()
    orig := notify
    calls := &[]notifyCall{}
    notify = func(title, msg string) {
        *calls = append(*calls, notifyCall{title: title, msg: msg})
    }
    t.Cleanup(func() { notify = orig })
    return calls
}
```

Each test then just calls `calls := captureNotify(t)` and removes the boilerplate.

**Rule Used:** What: Eliminate duplicate logic, functions, or cod... ([source](https://app.greptile.com/review/custom-context?memory=f0a796a9-684f-4dfb-a5cf-8c99c16b63a1))

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/utils/notify.go
Line: 28-34

Comment:
**Spurious blank line inside function body**

There is an empty line between the function signature and the first statement. This is non-idiomatic Go (gofmt/golangci-lint will flag it) and may cause the lint CI step to fail.

```suggestion
func Notify(title, message string) {
	err := beeep.Notify(title, message, iconPath)
```

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: a53a107</sub>

> Greptile also left **1 inline comment** on this PR.

**Context used:**

- Rule used - What: Eliminate duplicate logic, functions, or cod... ([source](https://app.greptile.com/review/custom-context?memory=f0a796a9-684f-4dfb-a5cf-8c99c16b63a1))
- Rule used - What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))

### CODE REVIEW

File: .github/workflows/binary-size-compare.yml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      - name: Comment on PR (Create or Update)
+        uses: actions/github-script@v7
+        with:
+          script: |
+            const pr = Number('${{ steps.compare.outputs.size_pr }}');
+            const main = Number('${{ steps.compare.outputs.size_main }}');
+            const diff = Number('${{ steps.compare.outputs.diff }}');
+            const exceeds = '${{ steps.compare.outputs.exceeds }}' === 'true';
+
+            // Better formatter handling negative numbers correctly
```
[the remark this answers is not in the document]
Comment (SuperCoolPencil) [reply .github/workflows/binary-size-compare.yml:69]: a developer would have to explicitly write a bug into the script like this
